text/0000-char-uax-31.md

```diff
@@ -0,0 +1,103 @@
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add functions to the standard library for testing a `char` against [UAX TR31](https://unicode.org/reports/tr31/) ("Unicode Annex 31")
+`Pattern_White_Space`, `Pattern_Syntax`, `XID_Start`, `ID_Nonstart`, and `XID_Continue` (the XID ones are already in the standard
+library, but are unstable; this RFC proposes to stablize them).
+
+# Motivation
+[motivation]: #motivation
+
+As a systems language, Rust is heavily used for parsing.
+As a progressive, forward-thinking language that accepts anyone,
+Rust supports Unicode and makes the definitive string types UTF-8.
+At the intersection of these needs sits *UAX #31: Unicode Identifier and Pattern Syntax* ("Annex 31"),
+a standardized set of code point categories for defining computer language syntax.
+
+This is being used in production Rust code already.
+Rust's own compiler already has functions to check against Annex 31 code point categories in the lexer,
+[but not everyone who works on the compiler knows about them](https://internals.rust-lang.org/t/for-await-loops/9819/16),
+and since they're not in the standard library,
+not everyone who works on Rust-related tooling has access to them.
+I'm not asserting that putting these in libstd would've avoided that bug,
+but if it was in the standard library,
+it would resolve the questions about whether third-party tooling can be expected to support the full range of Unicode whitespace.
+
+[Other languages](https://rosettacode.org/wiki/Unicode_variable_names#C) also follow Annex 31, such as C# and Elixir.
+Other common grammars, even ones that aren't actually for programming languages, can also be found or defined in Annex 31,
+such as hashtags and XML.
+
+It's also pretty clear what the "right" API is for this,
+since `is_whitespace` and `is_ascii_whitespace` already set the precedent here,
+so there's little need to experiment with API design.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+In addition to functions for checking "ASCII white space" and "Unicode white space,"
+some languages, such as Rust and C#, use Unicode Annex 31 to define their syntax.
+These functions are also exposed as methods on the `char` type.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## `fn char::is_id_nonstart(self) -> bool`
+
+Check if `self` is a member of Unicode Annex 31's `ID_Nonstart` code point category.
+This function is defined as `self.is_xid_continue() && !self.is_xid_start()`.
+
+## `fn char::is_pattern_syntax(self) -> bool`
+
+Check if `self` is a member of Unicode Annex 31's `Pattern_Syntax` code point category.
+
+## `fn char::is_pattern_white_space(self) -> bool`
+
+Check if `self` is a member of Unicode Annex 31's `Pattern_White_Space` code point category.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
```

Comment:
Matching unicode versions is also a big issue here. This function will be inaccurate half the time as we don't always update our data files immediately (it's not always straightforward). Even if we do, the behavior of this function will change every year, and while we don't have a guarantee on stdlib behavior stability, this does mean that older compilers will lead to different results on code that compiles. This further makes me feel like this should be a versioned crate.

is_whitespace already opened the doors to this issue, but I don't want to make it worse. is_whitespace is a small relatively stable list whereas XID expands all the time.

Reply:
Agreed. Added it.